google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/TestBench.java

```diff
@@ -0,0 +1,307 @@
+        try {
+          // wait a small amount of time for the server to come up before probing
+          Thread.sleep(500);
+          // wait for the server to come up
+          List<RetryTestResource> existingResources =
+              runWithRetries(
+                  TestBench.this::listRetryTests,
+                  RetrySettings.newBuilder()
+                      .setTotalTimeout(Duration.ofSeconds(30))
+                      .setInitialRetryDelay(Duration.ofMillis(500))
+                      .setRetryDelayMultiplier(1.5)
+                      .setMaxRetryDelay(Duration.ofSeconds(5))
+                      .build(),
+                  new BasicResultRetryAlgorithm<List<RetryTestResource>>() {
+                    @Override
+                    public boolean shouldRetry(
+                        Throwable previousThrowable, List<RetryTestResource> previousResponse) {
+                      return previousThrowable instanceof SocketException;
+                    }
+                  },
+                  NanoClock.getDefaultClock());
+          if (!existingResources.isEmpty()) {
+            LOGGER.info(
+                "Test Server already has retry tests in it, is it running outside the tests?");
+          }
+          base.evaluate();
+        } catch (RetryHelperException e) {
+          dumpServerLogs(outFile, errFile);
+          throw new IllegalStateException(
+              "Failed to connect to server within a reasonable amount of time. Host url: "
+                  + baseUri,
+              e.getCause());
+        } finally {
+          process.destroy();
+        }
+      }
+    };
+  }
+
+  private void dumpServerLogs(File outFile, File errFile) throws IOException {
+    try {
+      LOGGER.warning("Dumping contents of stdout");
+      dumpServerLog("stdout", outFile);
+    } finally {
+      LOGGER.warning("Dumping contents of stderr");
+      dumpServerLog("stderr", errFile);
+    }
+  }
+
+  private void dumpServerLog(String prefix, File out) throws IOException {
+    try (BufferedReader reader = new BufferedReader(new FileReader(out))) {
+      String line;
+      while ((line = reader.readLine()) != null) {
+        LOGGER.warning("<" + prefix + "> " + line);
+      }
+    }
+  }
+
+  static Builder newBuilder() {
+    return new Builder();
+  }
+
+  private void setCommonReqProperties(HttpRequest req) {
+    req.setCurlLoggingEnabled(false);
+    req.getHeaders().setAccept("application/json");
+    req.getHeaders().setUserAgent("java-conformance-tests/");
+  }
+
+  static final class RetryTestResource {
+    public String id;
+    public Boolean completed;
+    public JsonObject instructions;
+
+    @Override
+    public String toString() {
+      return "RetryTestResource{"
+          + "id='"
+          + id
+          + '\''
+          + ", completed="
+          + completed
+          + ", instructions="
+          + instructions
+          + '}';
+    }
+  }
+
+  static final class Builder {
```

Comment:
This is really interesting. I'd always envisioned the flow as something like 1. CI script spins up server then 2. CI script kicks off language specific test build. Instead, you are having the java test session itself drive launching the server. Seems like a fine approach but just wanted to call it out-- do you see any risks here?

Reply:
That's fair, with this approach I've ensured anyway someone can start the JUnit test they will be able to run (as long as they have docker) rather than having to coordinate running a server first manually before running the tests.

Reply:
Interesting way +1  Discovered package `subprocess` in python that can do something similar. It's working nice locally, but I haven't incorporated it with kokoro yet. Will we have trouble running the storage-testbench docker image within the kokoro trampoline docker container?

Reply:
The kokoro trampoline does all the necessary wiring so that docker containers can be ran from within the trampoline as well.

In fact, you can see the test results from one of the integration tests builds [here](https://source.cloud.google.com/results/invocations/2d9a7f9a-8c95-42cc-9707-1666013ff9e0/targets/github%2Fjava-storage%2Fgoogle-cloud-storage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.storage.conformance.retry.ITRetryConformanceTest/tests).